src/rgb.rs

```diff
@@ -0,0 +1,140 @@
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::slice;
+
+use libavif_sys as sys;
+
+use crate::{AvifImage, YuvFormat};
+
+pub struct RgbPixels<'a> {
+    owned: bool,
+    inner: sys::avifRGBImage,
+
+    phantom: PhantomData<&'a [u8]>,
+}
+
+impl<'a> RgbPixels<'a> {
+    pub fn new(width: u32, height: u32, rgb: &'a [u8]) -> Self {
+        let (stride, format) = if (width * height * 3) as usize == rgb.len() {
+            // RGB
+            (3, sys::AVIF_RGB_FORMAT_RGB)
+        } else if (width * height * 4) as usize == rgb.len() {
+            // RGBA
+            (4, sys::AVIF_RGB_FORMAT_RGBA)
+        } else {
+            panic!("invalid rgb len")
+        };
+
+        Self {
+            owned: true,
+            inner: sys::avifRGBImage {
+                width,
+                height,
+                depth: 8,
+                format,
+                chromaUpsampling: sys::AVIF_CHROMA_UPSAMPLING_BILINEAR,
+                ignoreAlpha: 0,
+                pixels: rgb.as_ptr() as *mut u8,
+                rowBytes: stride * width,
+            },
+            phantom: PhantomData,
+        }
+    }
+
+    /// width of the image in pixels
+    pub fn width(&self) -> u32 {
+        self.inner.width
+    }
+
+    /// height of the image in pixels
+    pub fn height(&self) -> u32 {
+        self.inner.height
+    }
+
+    pub fn pixel(&self, x: u32, y: u32) -> (u8, u8, u8, u8) {
+        let stride = if self.inner.format == sys::AVIF_RGB_FORMAT_RGBA {
+            4
+        } else {
+            3
+        };
+
+        let row_bytes = self.inner.rowBytes as usize;
+        let i = (stride * x as usize) + (row_bytes * y as usize);
+
+        let slice = self.as_slice();
+        let slice = &slice[i..][..stride];
+        (
+            slice[0],
+            slice[1],
+            slice[2],
+            if stride == 4 { slice[3] } else { 255 },
+        )
+    }
+
+    /// Extracts a slice containg all of the pixels without doing clones or allocation.
+    pub fn as_slice(&'a self) -> &'a [u8] {
+        let size = self.inner.rowBytes * self.height();
+        unsafe { slice::from_raw_parts(self.inner.pixels, size as usize) }
+    }
+
+    /// Converts `self` into a new vector by cloning all of the pixels.
+    pub fn to_vec(&self) -> Vec<u8> {
+        self.as_slice().to_vec()
+    }
+
+    pub fn to_image(&self, yuv_format: YuvFormat) -> AvifImage {
+        unsafe {
+            let image =
+                sys::avifImageCreate(self.width() as _, self.height() as _, 8, yuv_format as i32);
```

Comment:
nobody ever calls `avifImageDestroy` for this `avifImageCreate`, and I think there are more cases of that with other allocations.

Reply:
I forgot to write the Drop implementation 😅